### PR TITLE
Scanning a single resource

### PR DESCRIPTION
**Pull request by amirmalka**
## Overview

This PR adds an ability to run a scan on a single resource - work in progress.
Support has been added to the `K8sResourceHandler` and to the `FileResourceHandler` resource handlers.

The change mainly involves the scan engine. Once integration with other developments is completed, it will be possible to scan a single resource via the CLI or via the HTTP handler.

### List of changes:
- TBD

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,7 +19,7 @@ require (
 	github.com/kubescape/go-git-url v0.0.25
 	github.com/kubescape/go-logger v0.0.13
 	github.com/kubescape/k8s-interface v0.0.116
-	github.com/kubescape/opa-utils v0.0.253
+	github.com/kubescape/opa-utils v0.0.256-0.20230720131313-796d89cc623c
```
Comment (amirmalka) [inline go.mod:22]: https://github.com/kubescape/opa-utils/pull/127

File: cmd/scan/workload.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+				logger.L().Fatal(err.Error())
+			}
+
+			scanInfo.WorkloadIdentifier = &cautils.WorkloadIdentifier{
+				Namespace: namespace,
+				Kind:      kind,
+				Name:      name,
+			}
+
+			scanInfo.ScanAll = true
```
Comment (amirmalka) [inline cmd/scan/workload.go:68]: to be removed

File: core/pkg/opaprocessor/processorhandler.go [hunk 3]
```diff
@@ -227,6 +227,11 @@ func (opap *OPAProcessor) processControl(ctx context.Context, control *reporthan
 	allResources := make(map[string]workloadinterface.IMetadata, heuristicAllocResources)
 
 	for i := range control.Rules {
+		if opap.ExcludedRules != nil {
```
Comment (amirmalka) [inline core/pkg/opaprocessor/processorhandler.go:230]: we need to decide if we use this map to excluded irrelevant rules or change the Policies object in the first place (that will require a deep copy of the Policies object)

File: core/cautils/datastructures.go [hunk 4]
```diff
@@ -18,9 +18,10 @@ type K8SResources map[string][]string
 type KSResources map[string][]string
 
 type OPASessionObj struct {
-	K8SResources          *K8SResources                                 // input k8s objects
-	ArmoResource          *KSResources                                  // input ARMO objects
+	K8SResources          K8SResources                                  // input k8s objects
+	KubescapeResource     KSResources                                   // input Kubescape objects
```
Comment (amirmalka) [inline core/cautils/datastructures.go:22]: non-k8s object

File: core/pkg/resourcehandler/k8sresources.go [hunk 5]
```diff
@@ -153,10 +157,56 @@ func (k8sHandler *K8sResourceHandler) GetResources(ctx context.Context, sessionO
...
+	gvr, err := k8sinterface.GetGroupVersionResource(wlIdentifierString)
+	if err != nil {
+		return nil, err
+	}
+
+	fieldSelectors := getNameFieldSelectorString(resource.GetName(), "=")
+	if resource.GetNamespace() != "" && k8sinterface.IsNamespaceScope(&gvr) {
+		fieldSelectors = combineFieldSelectors(fieldSelectors, getNamespaceFieldSelectorString(resource.GetNamespace(), "="))
+	}
+	result, err := k8sHandler.pullSingleResource(&gvr, nil, fieldSelectors)
```
Comment (amirmalka) [inline core/pkg/resourcehandler/k8sresources.go:187]: confirm that only single resource returns

File: core/pkg/resourcehandler/resourcehandlerutils.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+				if resourcesFilterMap != nil {
+					if relevant := resourcesFilterMap[resource]; !relevant {
+						continue
+					}
+				}
+
+				groupResources := k8sinterface.ResourceGroupToString(apiGroup, apiVersions, resource)
+				// if namespace filter is set, we are scanning a workload in a specific namespace
+				// calling the getNamespacesSelector will add the namespace field selector (or name for Namespace resource)
+				globalFieldSelector := getNamespacesSelector(resource, namespace, "=")
```
Comment (amirmalka) [inline core/pkg/resourcehandler/resourcehandlerutils.go:125]: `--include-namespaces` && `--excluded-namespaces` check if broken

File: core/pkg/resourcehandler/queryableresource.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	GroupVersionResourceTriplet string
+	// metadata.name==<resource name>, metadata.namespace==<resource namespace> etc.
+	FieldSelectors string
+}
+
+func (qr *QueryableResource) String() string {
+	if qr.FieldSelectors == "" {
+		return qr.GroupVersionResourceTriplet
+	}
+	return fmt.Sprintf("%s/%s", qr.GroupVersionResourceTriplet, qr.FieldSelectors)
```
Comment (amirmalka) [inline core/pkg/resourcehandler/queryableresource.go:23]: fmt.Sprintf -> better use string.join

**Review by dwertent — APPROVED** [review]

File: core/pkg/resourcehandler/resourcehandlerutils.go (hunk 6 above)
Comment (amirmalka) [reply core/pkg/resourcehandler/resourcehandlerutils.go:125]: @dwertent flags are working

**Pull request merged**